File: Strategy/typeScript/strategy.ts
interface IOperation {
  calculate(number1: number, number2: number): number;
}

class Sum implements IOperation {
  calculate(number1: number, number2: number): number {
    console.log('add');
    return number1 + number2;
  }
}

class Diff implements IOperation {
  calculate(number1: number, number2: number): number {
    console.log('diff');
    return number1 - number2;
  }
}

class Multi implements IOperation {
  calculate(number1: number, number2: number): number {
    console.log('multi');
    return number1 * number2;
  }
}

class Divide implements IOperation {
  calculate(number1: number, number2: number): number {
    console.log('div');
    return number1 / number2;
  }
}

class MathOperations {
  private _instance: IOperation;
  constructor(instance: IOperation) {
    this._instance = instance;
  }

  public set instance(value: IOperation) {
    this._instance = value;
  }

  execute(number1: number, number2: number) {
    return this._instance.calculate(number1, number2);
  }
}

const mathOperations = new MathOperations(new Sum());
console.log(mathOperations.execute(1, 2));
mathOperations.instance = new Diff();
console.log(mathOperations.execute(1, 2));
mathOperations.instance = new Multi();
console.log(mathOperations.execute(4, 2));
mathOperations.instance = new Divide();
console.log(mathOperations.execute(3, 2));

File: Strategy/python/strategy.py
from abc import ABC, abstractmethod


class Operation(ABC):
    @abstractmethod
    def calculate(self, number1: float, number2: float) -> float:
        pass


class Sum(Operation):

    def calculate(self, number1: float, number2: float) -> float:
        return number1 + number2


class Rest(Operation):

    def calculate(self, number1: float, number2: float) -> float:
        return number1 - number2


class MathOperation(ABC):

    def __init__(self, instance: Operation):
        self.__instance = instance

    @property
    def instance(self):
        return self.__instance

    @instance.setter
    def instance(self, instance: Operation):
        self.__instance = instance

    def calculate(self, number1: float, number2: float) -> float:
        return self.__instance.calculate(number1, number2)


if __name__ == '__main__':
    mathOperators = MathOperation(Sum())
    print(mathOperators.instance)
    print(mathOperators.calculate(1, 2))
    mathOperators.instance = Rest()
    print(mathOperators.instance)
    print(mathOperators.calculate(1, 2))
